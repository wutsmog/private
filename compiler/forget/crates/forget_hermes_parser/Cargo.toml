[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }