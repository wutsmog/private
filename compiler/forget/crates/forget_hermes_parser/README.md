# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.