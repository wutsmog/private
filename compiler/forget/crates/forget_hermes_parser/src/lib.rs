mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
