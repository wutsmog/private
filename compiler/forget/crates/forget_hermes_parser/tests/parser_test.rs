use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
