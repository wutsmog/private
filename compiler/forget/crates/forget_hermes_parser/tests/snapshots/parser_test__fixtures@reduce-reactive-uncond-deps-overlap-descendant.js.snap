---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/reduce-reactive-uncond-deps-overlap-descendant.js
---
Input:
// Test that we correctly track a subpath if the subpath itself is accessed as
// a dependency
function TestOverlappingDescendantTracked(props) {
  let x = {};
  x.b = props.a.b.c;
  x.c = props.a.b.c.x.y;
  x.a = props.a;
  return x;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "TestOverlappingDescendantTracked",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 104,
          "end": 136
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 137,
            "end": 142
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "let",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 152,
                    "end": 153
                  }
                },
                "init": {
                  "type": "ObjectExpression",
                  "properties": [],
                  "loc": null,
                  "range": {
                    "start": 156,
                    "end": 158
                  }
                },
                "loc": null,
                "range": {
                  "start": 152,
                  "end": 158
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 148,
              "end": 159
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "AssignmentExpression",
              "operator": "=",
              "left": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 162,
                    "end": 163
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "b",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 164,
                    "end": 165
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 162,
                  "end": 165
                }
              },
              "right": {
                "type": "MemberExpression",
                "object": {
                  "type": "MemberExpression",
                  "object": {
                    "type": "MemberExpression",
                    "object": {
                      "type": "Identifier",
                      "name": "props",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 168,
                        "end": 173
                      }
                    },
                    "property": {
                      "type": "Identifier",
                      "name": "a",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 174,
                        "end": 175
                      }
                    },
                    "computed": false,
                    "loc": null,
                    "range": {
                      "start": 168,
                      "end": 175
                    }
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "b",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 176,
                      "end": 177
                    }
                  },
                  "computed": false,
                  "loc": null,
                  "range": {
                    "start": 168,
                    "end": 177
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "c",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 178,
                    "end": 179
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 168,
                  "end": 179
                }
              },
              "loc": null,
              "range": {
                "start": 162,
                "end": 179
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 162,
              "end": 180
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "AssignmentExpression",
              "operator": "=",
              "left": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 183,
                    "end": 184
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "c",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 185,
                    "end": 186
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 183,
                  "end": 186
                }
              },
              "right": {
                "type": "MemberExpression",
                "object": {
                  "type": "MemberExpression",
                  "object": {
                    "type": "MemberExpression",
                    "object": {
                      "type": "MemberExpression",
                      "object": {
                        "type": "MemberExpression",
                        "object": {
                          "type": "Identifier",
                          "name": "props",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 189,
                            "end": 194
                          }
                        },
                        "property": {
                          "type": "Identifier",
                          "name": "a",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 195,
                            "end": 196
                          }
                        },
                        "computed": false,
                        "loc": null,
                        "range": {
                          "start": 189,
                          "end": 196
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "b",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 197,
                          "end": 198
                        }
                      },
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 189,
                        "end": 198
                      }
                    },
                    "property": {
                      "type": "Identifier",
                      "name": "c",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 199,
                        "end": 200
                      }
                    },
                    "computed": false,
                    "loc": null,
                    "range": {
                      "start": 189,
                      "end": 200
                    }
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "x",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 201,
                      "end": 202
                    }
                  },
                  "computed": false,
                  "loc": null,
                  "range": {
                    "start": 189,
                    "end": 202
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 203,
                    "end": 204
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 189,
                  "end": 204
                }
              },
              "loc": null,
              "range": {
                "start": 183,
                "end": 204
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 183,
              "end": 205
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "AssignmentExpression",
              "operator": "=",
              "left": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 208,
                    "end": 209
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "a",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 210,
                    "end": 211
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 208,
                  "end": 211
                }
              },
              "right": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "props",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 214,
                    "end": 219
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "a",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 220,
                    "end": 221
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 214,
                  "end": 221
                }
              },
              "loc": null,
              "range": {
                "start": 208,
                "end": 221
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 208,
              "end": 222
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "Identifier",
              "name": "x",
              "typeAnnotation": null,
              "loc": null,
              "range": {
                "start": 232,
                "end": 233
              }
            },
            "loc": null,
            "range": {
              "start": 225,
              "end": 234
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 144,
          "end": 236
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 95,
        "end": 236
      },
      "loc": null,
      "range": {
        "start": 95,
        "end": 236
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 95,
    "end": 236
  }
}
